rascal-lsp/src/main/rascal/lang/rascal/lsp/Actions.rsc

```diff
@@ -0,0 +1,106 @@
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+}
+@bootstrapParser
+@synopsis{Defines both the command evaluator and the codeAction retriever for Rascal}
+module lang::rascal::lsp::Actions
+
+import lang::rascal::\syntax::Rascal;
+import util::LanguageServer;
+import analysis::diff::edits::TextEdits;
+import ParseTree;
+import String;
+import lang::rascal::vis::ImportGraph;
+import util::Reflective;
+import util::IDEServices;
+import List;
+
+@synopsis{Here we list Rascal-specific code commands}
+@description{
+The commands must be evaluated by ((evaluateRascalCommand))
+}
+data Command
+    = visualImportGraphCommand(PathConfig pcfg)
+    | sortImportsAndExtends(Header h)
+    ;
+
+@synopsis{Detects (on-demand) source actions to register with specific places near the current cursor}
+list[CodeAction] rascalCodeActions(Focus focus, PathConfig pcfg=pathConfig()) {
+    result = [];
+
+    if ([*_, Toplevel t, *_] := focus) {
+        result += toplevelCodeActions(t);
+    }
+
+    if ([*_, Header h, *_] := focus) {
+        result += [action(command=visualImportGraphCommand(pcfg), title="Visualize project import graph", kind=source())]
+               +  [action(command=sortImportsAndExtends(h), title="Sort imports and extends", kind=source())]
+               ;
+    }
+
+    return result;
+}
+
+@synopsis{Rewrite immediate return to expression.}
+list[CodeAction] toplevelCodeActions(Toplevel t:
+    (Toplevel) `<Tags tags>
+               '<Visibility visibility> <Signature signature> {
+               '  return <Expression e>;
+               '}`) {
+
+    result = (Toplevel) `<Tags tags>
+                        '<Visibility visibility> <Signature signature> = <Expression e>;`;
+
+    edits=[changed(t@\loc.top, [replace(t@\loc, trim("<result>"))])];
+
+    return [action(edits=edits, title="Rewrite block return to simpler rewrite rule.", kind=refactor())];
+}
+
+default list[CodeAction] toplevelCodeActions(Toplevel _) = [];
+
+@synopsis{Evaluates all commands and quickfixes produced by ((rascalCodeActions)) and the type-checker}
+default value evaluateRascalCommand(Command _) =  ("result" : false);
+
+value evaluateRascalCommand(visualImportGraphCommand(PathConfig pcfg)) {
+    importGraph(pcfg);
+    return ("result" : true);
+}
+
+value evaluateRascalCommand(sortImportsAndExtends(Header h)) {
+    extends = [trim("<i>") | i <- h.imports, i is \extend];
+    imports = [trim("<i>") | i <- h.imports, i is \default];
+    grammar = [trim("<i>") | i <- h.imports, i is \syntax];
+
+    newHeader = "<for (i <- sort(extends)) {><i>
+                '<}>
+                '<for (i <- sort(imports)) {><i>
+                '<}>
+                '<for (i <- grammar) {><i>
+                '<}>";
+
+    applyDocumentsEdits([changed(h@\loc.top, [replace(h.imports@\loc, newHeader)])]);
```

Comment:
Why does this directly call `applyDocumentEdits` while the top-level rewrite on line 79 returns an `action(..., kind = refactor())`?

Reply:
It's a demo of the possibilities. There is no semantic difference other than a different time of synchronization. The user does not see a difference. Both ways are applied in the editor.

Reply:
However, there are reasons for choosing a command over edits, typically if the amount of edits is very large or expensive to compute, then the action edits will block the creation of the actions menu. In that case it is always better to register a command that will compute the edits later.

Reply:
other reasons are if you need other side-effects next to the edits, then a command is useful. If you pick both, say an action with both edits and a command, then the edits will always be executed before the command is executed.